File: projects/basis/sdk-review/py-v2/basis/modules/leverage_simulator.py
from web3 import Web3
from .factory import load_abi


class LeverageSimulatorModule:
    def __init__(self, client, leverage_address: str):
        self.client = client
        self.leverage_address = Web3.to_checksum_address(leverage_address)
        self.leverage_abi = load_abi('ALEVERAGE.json')
        self.contract = self.client.web3.eth.contract(address=self.leverage_address, abi=self.leverage_abi)

    # ------------------------------------------------------------------
    # Read methods
    # ------------------------------------------------------------------

    def simulate_leverage(self, amount: int, path: list[str], number_of_days: int):
        """Simulates a leverage position and returns the result."""
        checksum_path = [Web3.to_checksum_address(p) for p in path]
        return self.contract.functions.simulateLeverage(amount, checksum_path, number_of_days).call()

    def simulate_leverage_factory(self, amount: int, path: list[str], number_of_days: int):
        """Simulates a leverage position for factory tokens and returns the result."""
        checksum_path = [Web3.to_checksum_address(p) for p in path]
        return self.contract.functions.simulateLeverageFactory(amount, checksum_path, number_of_days).call()

    def calculate_floor(self, hybrid_multiplier: int, reserve0: int, reserve1: int, base_reserve0: int, xe_reserve0: int, xe_reserve1: int) -> int:
        """Calculates the floor price for a hybrid token."""
        return self.contract.functions.calculateFloor(hybrid_multiplier, reserve0, reserve1, base_reserve0, xe_reserve0, xe_reserve1).call()

    def get_token_price(self, reserve0: int, reserve1: int) -> int:
        """Returns the token price given reserves."""
        return self.contract.functions.getTokenPrice(reserve0, reserve1).call()

    def get_usd_price(self, reserve0: int, reserve1: int, xe_reserve0: int, xe_reserve1: int) -> int:
        """Returns the USD price given reserves."""
        return self.contract.functions.getUSDPrice(reserve0, reserve1, xe_reserve0, xe_reserve1).call()

    def get_collateral_value(self, token_amount: int, reserve0: int, reserve1: int) -> int:
        """Returns the collateral value for a given token amount."""
        return self.contract.functions.getColleteralValue(token_amount, reserve0, reserve1).call()

    def get_collateral_value_hybrid(self, token_amount: int, reserve0: int, reserve1: int, xe_reserve0: int, xe_reserve1: int, multiplier: int, base_reserve0: int) -> int:
        """Returns the collateral value for a hybrid token."""
        return self.contract.functions.getColleteralValueHybrid(token_amount, reserve0, reserve1, xe_reserve0, xe_reserve1, multiplier, base_reserve0).call()

    def calculate_tokens_for_buy(self, usdb_amount: int, reserve0: int, reserve1: int) -> int:
        """Calculates the number of tokens received for a given USDB input."""
        return self.contract.functions.calculateTokensForBuy(usdb_amount, reserve0, reserve1).call()

    def calculate_tokens_to_burn(self, amount_in: int, multiplier: int, input_reserve0: int, input_reserve1: int, splitter: int) -> int:
        """Calculates the number of tokens to burn."""
        return self.contract.functions.calculateTokensToBurn(amount_in, multiplier, input_reserve0, input_reserve1, splitter).call()

File: projects/basis/sdk-review/py-v2/src/modules/LeverageSimulator.ts
import { BasisClient } from '../BasisClient';
import ALeverageArtifact from '../abis/ALEVERAGE.json';
import { Address } from 'viem';

export class LeverageSimulatorModule {
  private client: BasisClient;
  private leverageAddress: Address;

  constructor(client: BasisClient, leverageAddress: Address) {
    this.client = client;
    this.leverageAddress = leverageAddress;
  }

  /**
   * Simulates a leveraged buy and returns the EndResult struct.
   */
  async simulateLeverage(amount: bigint, path: Address[], numberOfDays: bigint) {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'simulateLeverage',
      args: [amount, path, numberOfDays],
    });
  }

  /**
   * Simulates a leveraged buy via factory and returns the EndResult struct.
   */
  async simulateLeverageFactory(amount: bigint, path: Address[], numberOfDays: bigint) {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'simulateLeverageFactory',
      args: [amount, path, numberOfDays],
    });
  }

  /**
   * Calculates the floor price for a hybrid token.
   */
  async calculateFloor(
    hybridMultiplier: bigint,
    reserve0: bigint,
    reserve1: bigint,
    baseReserve0: bigint,
    xereserve0: bigint,
    xereserve1: bigint
  ): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'calculateFloor',
      args: [hybridMultiplier, reserve0, reserve1, baseReserve0, xereserve0, xereserve1],
    }) as Promise<bigint>;
  }

  /**
   * Returns the token price given reserves.
   */
  async getTokenPrice(reserve0: bigint, reserve1: bigint): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'getTokenPrice',
      args: [reserve0, reserve1],
    }) as Promise<bigint>;
  }

  /**
   * Returns the USD price of a token given reserves.
   */
  async getUSDPrice(reserve0: bigint, reserve1: bigint, xereserve0: bigint, xereserve1: bigint): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'getUSDPrice',
      args: [reserve0, reserve1, xereserve0, xereserve1],
    }) as Promise<bigint>;
  }

  /**
   * Returns the collateral value in USDB for a given token amount.
   */
  async getCollateralValue(tokenAmount: bigint, reserve0: bigint, reserve1: bigint): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'getColleteralValue',
      args: [tokenAmount, reserve0, reserve1],
    }) as Promise<bigint>;
  }

  /**
   * Returns the collateral value for a hybrid token.
   */
  async getCollateralValueHybrid(
    tokenAmount: bigint,
    reserve0: bigint,
    reserve1: bigint,
    xereserve0: bigint,
    xereserve1: bigint,
    multiplier: bigint,
    basereserve0: bigint
  ): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'getColleteralValueHybrid',
      args: [tokenAmount, reserve0, reserve1, xereserve0, xereserve1, multiplier, basereserve0],
    }) as Promise<bigint>;
  }

  /**
   * Calculates how many tokens can be purchased for a given USDB amount.
   */
  async calculateTokensForBuy(usdbAmount: bigint, reserve0: bigint, reserve1: bigint): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'calculateTokensForBuy',
      args: [usdbAmount, reserve0, reserve1],
    }) as Promise<bigint>;
  }

  /**
   * Calculates the number of tokens to burn for a given input.
   */
  async calculateTokensToBurn(
    amountIn: bigint,
    multiplier: bigint,
    inputreserve0: bigint,
    inputreserve1: bigint,
    splitter: bigint
  ): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.leverageAddress,
      abi: ALeverageArtifact.abi,
      functionName: 'calculateTokensToBurn',
      args: [amountIn, multiplier, inputreserve0, inputreserve1, splitter],
    }) as Promise<bigint>;
  }
}
